Add -AsHashtable switch to ConvertFrom-Json for issue #3623

bergmeister opened the pull request:
Address https://github.com/PowerShell/PowerShell/issues/3623.

Implementation symmetric to traditional code path that returns a `PSCustomObject`. Code is shared on the top level but 2 low level methods were difficult to share, therefore 2 separate but similar methods were created for HashTables.
All existing tests related to this change were adapted to also test against this new switch and were slightly improved. 

Update (@TravisEz13):
`-AsHashtable` is an existing pattern used in `Group-Object`

bergmeister committed:
Add -AsHashtable switch to ConvertFrom-Json for issue 3623

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -26,13 +27,14 @@ public static class JsonObject
+        /// <returns>A PSObject or a Hashtable if the <paramref name="returnHashTable"/> parameter is true.</returns>
         [SuppressMessage("Microsoft.Naming", "CA1704:IdentifiersShouldBeSpelledCorrectly")]
-        public static object ConvertFromJson(string input, out ErrorRecord error)
+        public static object ConvertFromJson(string input, bool returnHashTable, out ErrorRecord error)
```
We shouldn't change a public API. Please add new overload.

bergmeister replied:
Thanks. I did not know PowerShell also exposes some of its classes. I'll re-add the old signature API  that will then internally call the new overload.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,33 @@
+        @{ name = "empty";  str = "";      ReturnHashTable = $false }
+        @{ name = "spaces"; str = "  ";    ReturnHashTable = $false }
+        @{ name = "object"; str = "{a:1}"; ReturnHashTable = $false }
     )
```
Please move to BeforeAll block.

bergmeister replied:
I am not sure what you exactly mean by that. I removed the redundant line about `$TestCasesForReturnHashTableParameter that was a leftover.

iSazonov replied:
Please see [sample](https://github.com/PowerShell/PowerShell/blob/309bf603f9aff9009dce7e725d42b2d4f99f58c9/test/powershell/Modules/Microsoft.PowerShell.Management/Clear-Item.Tests.ps1#L2) 
We should place ` $TestCasesForReturnHashTableParameter ` and `$validStrings ` in the BeforeAll block.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,33 @@
 
-    It 'no error for valid string - <name>' -TestCase $validStrings {
-        param ($str)
+    It 'no error for valid string - <name> when ReturnHashTable is <ReturnHashTable>' -TestCase $validStrings {
```
Maybe 'no error for valid string - <name> with -ReturnHashTable:`$\<ReturnHashTable>'

bergmeister replied:
I changed to match your suggestion whilst still displaying the name of the string.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/ConvertFromJsonCommand.cs:
```diff
@@ -20,17 +20,23 @@ public class ConvertFromJsonCommand : Cmdlet
+        /// <summary>
+        /// Returned data structure is a Hashtable instead a CustomPSObject.
+        /// </summary>
+        [Parameter(Mandatory = false)]
```
It is default. Please remove.

bergmeister replied:
What do you mean by that? Can you elaborate please (what is 'it')?

bergmeister replied:
Ok. I see, thanks.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs:
```diff
@@ -65,15 +67,29 @@ public static object ConvertFromJson(string input, out ErrorRecord error)
+                        {
+                            obj = PopulateFromJArray(list, out error);
+                        }
                     }
```
Please add 'else' with Diagnostics.Assert (if the object is not JObject or JArray).

bergmeister replied:
Because of the cast to `JArray` a few lines above, `obj` cannot be something else. I initially put your suggestion in but even the compiler tells me that:
`error CS0184: The given expression is never of the provided ('JObject') type`

iSazonov replied:
I meant:
```c#
                        Dbg.Diagnostics.Assert(
                            false,
                            "Only JObject or JArray is supported.");

```

iSazonov reviewed (commented):
@bergmeister Thanks for great contribution!

bergmeister committed:
Keep old ConvertFromJson API for backwards compatibility and address minor suggestions in test naming

bergmeister committed:
Remove redundant '[Parameter(Mandatory = false)]' on -AsHashtable switch since the parameter is not mandatory by default

bergmeister committed:
Last commit accidentally removed entirely [Parameter()] line, it should have been only the Mandatory=false part that should be removed.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/Common/InvokeRestMethodCommand.Common.cs:
```diff
@@ -171,7 +171,7 @@ private bool TryConvertToJson(string json, out object obj, ref Exception exRef)
             {
                 ErrorRecord error;
-                obj = JsonObject.ConvertFromJson(json, out error);
+                obj = JsonObject.ConvertFromJson(json, false, out error);
```
Please revert the change - we have the overload.

bergmeister replied:
Yes, you're right.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,31 @@
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $true  }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $true  }
+        @{ name = "plain text"; str = "plaintext"; ReturnHashTable = $false }
+        @{ name = "part";       str = '{"a" :';    ReturnHashTable = $false }
```
I see we have a check for `input == null` but haven't the test - please add new test with `str = $null`.

bergmeister replied:
OK. Added

bergmeister committed:
-Revert change in InvokeRestMethodCommand originally made in this PR that is not necessary any more since the original method signature has been restored.
-Improve test to add case for $null

iSazonov commented:
We have discussion for this https://github.com/PowerShell/PowerShell/issues/3623

SteveL-MSFT commented:
As part of this PR, if we can detect that the json has an property named empty string for the pscustomobject path and provide an error message like:

```markdown
The provided json includes a property who name is an empty string, this is only supported using the `-AsHashTable` switch.
```

Then we can resolve https://github.com/PowerShell/PowerShell/issues/1755 as fixed.

bergmeister committed:
Detect if key is empty string to resolve 1755 issue by redirecting to -AsHashTable via error message. Fixed a typo that I found in the message strings.

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:
```diff
@@ -126,9 +126,12 @@
     <value>Path '{0}' resolves to a directory. Specify a path including a file name, and then retry the command.</value>
   </data>
+  <data name="EmptyKeyInJsonString" xml:space="preserve">
+    <value>The provided json includes a property whose name is an empty string, this is only supported using the -AsHashTable switch.</value>
```
Should probably capitalize json as `JSON`

bergmeister replied:
OK. Fixed it.

bergmeister committed:
Fix test syntax and simplify test name

bergmeister committed:
make 'json' uppercase 'JSON' in error message as suggested in PR review.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,47 @@
 
-    It 'no error for valid string - <name>' -TestCase $validStrings {
-        param ($str)
+    It 'no error for valid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase $validStrings {
```
Please start test descriptions with capital.

We use `$validStrings` one time. Please use our common template:
```powershell
It 'no error for valid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase @(
 @{ name = "null";   str = $null;   ReturnHashTable = $true  }
 ...
) {
...
}
```

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,47 @@
 
-    It 'throw ArgumentException for invalid string - <name>' -TestCase $invalidStrings  {
-        param ($str)
+    It 'throw ArgumentException for invalid string ''<name>'' with -ReturnHashTable:$<ReturnHashTable>' -TestCase $invalidStrings  {
```
The same.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,47 @@
+        { [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, $ReturnHashTable, [ref]$errRecord) } | ShouldBeErrorId "ArgumentException"
+    }
+
+    $jsonWithEmptyKey = '{"": "Value"}'
```
We should put variables in `BeforeAll` block or in `It` block.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,47 @@
+    It 'throw InvalidOperationException when json contains empty key name' {
+        $errorRecord = $null
+        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, [ref]$errorRecord)
+        $errorRecord.Exception.GetType() | Should Be System.InvalidOperationException
```
Usually we don't check an exception types. Please remove.
(For type check we use Should BeOfType)

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,47 @@
+        $errorRecord = $null
+        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, [ref]$errorRecord)
+        $errorRecord.Exception.GetType() | Should Be System.InvalidOperationException
+        $errorRecord.FullyQualifiedErrorId | Should Be EmptyKeyInJsonString
```
Please put `EmptyKeyInJsonString` in quotas. It is more readable.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1:
```diff
@@ -1,16 +1,33 @@
-    It 'can convert a single-line object' {
-        ('{"a" : "1"}' | ConvertFrom-Json).a | Should Be 1
+    
+    $testCasesWithAndWithoutAsHashtableSwitch = @(
```
Please put this in `BeforeAll` block.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertFrom-Json.Tests.ps1:
```diff
@@ -1,16 +1,33 @@
+        @{ AsHashtable = $false }
+    )
+
+    It 'can convert a single-line object with AsHashtable switch set to <AsHashtable>' -TestCase $testCasesWithAndWithoutAsHashtableSwitch {
```
Please start test descriptions with capital.

iSazonov reviewed (commented):
Do we address the [comment](https://github.com/PowerShell/PowerShell/issues/3623#issuecomment-296537316) about case sensitivity?

bergmeister committed:
Address PR comments about tests (mostly style) by @iSazonov

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,6 +1,11 @@
+        $jsonWithEmptyKey = '{"": "Value"}'
+    }
+
+
```
Please remove extra line.

iSazonov commented:
@SteveL-MSFT @TravisEz13 @dantraMSFT Could you please review?

Also should we address the [comment](https://github.com/PowerShell/PowerShell/issues/3623#issuecomment-296537316) from @powercode 
>Apart from object size and performance, it may matter with case sensitivity. I have run into issues where I couldn't use the PowerShell cmdlets because the json contained 'Property' and 'property'.

bergmeister committed:
Update JsonObject.Tests.ps1

Removed extra newline in pester test as requested by @iSazonov

iSazonov commented:
@bergmeister Thanks for your contribution! LGTM with above question about  case sensitivity.

SteveL-MSFT commented:
@iSazonov @powercode regarding case sensitivity, I think we should treat it the same as the empty key.  [Hashtable] allows two keys differing in case, so if we detect that, we should just tell the user to use `-AsHashTable`.

bergmeister committed:
Report error when keys have different casing. As part of this I also found out that newtonsoft.json would squash keys that have the same casing into one and just use the last value. I still handle this case should the behaviour change.

bergmeister commented:
@SteveL-MSFT @iSazonov @powercode I added a specific error message to point the user to the new `-AsHashTable` option.
As part of this I also found out that actually no error is thrown in the case of duplicate keys with the same casing. This is due to the behaviour of Newtonsoft.Json that just uses the last item if there are multiple items with the same key of the same casing. The PowerShell code currently caters for it should this behaviour change in a newer version of Newtonsoft.Json. This should definitely be discussed in another issue if you want to make a breaking change to throw when encountering case matching duplicate keys.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/resources/WebCmdletStrings.resx:
```diff
@@ -133,7 +133,10 @@
+    <value>Cannot convert the JSON string because a dictionary that was converted from the string contains the duplicated key '{0}'.</value>
+  </data>
+  <data name="KeysWithDifferentCasingInJsonString" xml:space="preserve">
+    <value>Cannot convert the JSON string because a PSObject does not support keys with different casing. Please use the -AsHashTable switch instead. The key that was attempted to be added to the exisiting key '{0}' was '{1}'.</value>
```
Oh, very long message!

bergmeister replied:
I know. I shortened it a bit but I think the last sentence still adds value.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,63 @@
+        It 'Not throw when json contains key (same casing) when ReturnHashTable is true' {
+            $errorRecord = $null
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonContainingKeysWithDifferentCasing, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
```
I think it makes sense to check the hashtable values.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,63 @@
+        param ($str, $ReturnHashTable)
         $errRecord = $null
-        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, [ref]$errRecord)
+        [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($str, $ReturnHashTable, [ref]$errRecord)
```
Please validate the returned object

SteveL-MSFT replied:
You should capture the output as $result and validate it is what you expect

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,63 @@
+        It 'Not throw when json contains empty key name when ReturnHashTable is true' {
+            $errorRecord = $null
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
```
Please validate the returned object

SteveL-MSFT replied:
Resolved

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,63 @@
+        It 'Not throw when json contains key (same casing) when ReturnHashTable is true' {
+            $errorRecord = $null
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonContainingKeysWithDifferentCasing, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
```
Please validate the returned object

SteveL-MSFT replied:
This one is resolved.  Thanks.

iSazonov reviewed (commented):
@bergmeister Thanks for your contribution! I believe we are near to merge.
Feel free to open new Issue to discuss the discovered problem.

bergmeister committed:
Add more validation to positive tests and shorten error message as requested in PR.
Unfortunately, Pester (3.4) does not seem to support equality assertion for hashtables...

bergmeister commented:
@iSazonov @SteveL-MSFT I added more validation to the positive test cases and shortened the error message a bit. I opened issue #5199 for the problem with Newtonsoft.Json for keys with matching cases, which is unrelated to this PR.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/JsonObject.Tests.ps1:
```diff
@@ -1,46 +1,68 @@
+            $result = [Microsoft.PowerShell.Commands.JsonObject]::ConvertFromJson($jsonWithEmptyKey, $true, [ref]$errorRecord)
+            $result | Should Not Be $null
+            $result.Count | Should Be 1
+            $result.$('') | Should Be 'Value'
```
You should simplify this to just: $result."" | Should Be 'Value'

bergmeister committed:
Minor syntax simplification in jsonobject.tests.ps1

bergmeister commented:
@SteveL-MSFT I see that the documentation [here](https://github.com/PowerShell/PowerShell-Docs/blob/a1520688fb6b3e25ec51096d2611de9b54d08c2d/reference/6/Microsoft.PowerShell.Utility/ConvertFrom-Json.md) has not been updated yet for the new `-AsHashTable` switch (or is it already done in another branch?). Am I supposed to do that or shall I simply open an issue in the docs repo?

iSazonov commented:
@bergmeister Feel free to open an Issue in docs repo or push PR with the fix.

SteveL-MSFT commented:
@bergmeister preferably submit a PR to powershell-docs repo.  The `Documentation Needed` label here means you don't need to open an issue in the powershell-docs repo.  But my team won't be reviewing and adding docs until sometime after RC is out.  If you submit a PR for the doc update, please reference this PR so we can remove the `Documentation Needed` label.  Thanks!
